### PR TITLE
realtek: add support for L2 port isolation

### PR DESCRIPTION
**Pull request by sudoBash418**
This PR adds support for L2 port isolation to the `rtl83xx` DSA driver.

With this PR, enabling port isolation on two ports within the same bridge (via LuCI or otherwise) should result in no traffic being forwarded between the two ports.

I've tested these changes on my Zyxel GS1900-24E (RTL8382M).  
Testing on other devices (especially other SoCs) would be greatly appreciated :)

---

Questions:
- Does this work correctly on RTL931X devices?
I'm not sure if the `BUG: This does not work on RTL931X` comment is still accurate; if so, what exactly does not work? (I'd like to make the comment more specific at least)
- Is `6xx` the correct patch number for the kernel patch?

### CODE REVIEW

**Review by svanheule — COMMENTED** [review]
Any objection to using your name in the SoB?

**Comment by sudoBash418** [comment]
> Any objection to using your name in the SoB?

I'd prefer to use my username: it should quality as a known identity, and I'm more comfortable using it than my real name.
I did see the CI trip over it, but if I understand the [submission guidelines][1] correctly, a persistent screen name should be fine.

FWIW, it's also more unique (and also a minor hassle to change now that I've been authoring commits as `sudoBash418` for some time, but that's besides the point :smile:).

[1]: https://openwrt.org/submitting-patches#submission_guidelines

File: target/linux/realtek/files-5.15/drivers/net/dsa/rtl83xx/dsa.c [hunk 1]
```diff
@@ -1177,35 +1232,27 @@ static void rtl83xx_port_bridge_leave(struct dsa_switch *ds, int port,
...
+			/* ignore ports that are not members of the same bridge */
 			if (dsa_to_port(ds, i)->bridge_dev != bridge)
 				continue;
-			if (priv->ports[i].enable)
-				priv->r->traffic_disable(i, port);
 
-			priv->ports[i].pm &= ~BIT_ULL(port);
-			port_bitmap |= BIT_ULL(i);
+			/* remove ourselves from neighbor's bitmask */
+			priv->ports[i].bridge_neighbors |= BIT_ULL(port);
```
[the remark this answers is not in the document]
Comment (sudoBash418) [reply target/linux/realtek/files-5.15/drivers/net/dsa/rtl83xx/dsa.c:1252]: Yep, thanks for catching that :sweat_smile:

File: target/linux/realtek/files-5.15/drivers/net/dsa/rtl83xx/dsa.c [hunk 2]
```diff
@@ -1260,6 +1265,44 @@ static void rtl83xx_port_bridge_leave(struct dsa_switch *ds, int port,
...
+	if (port >= priv->cpu_port)
+		return -EOPNOTSUPP;
+
+	mutex_lock(&priv->reg_mutex);
+
+	/* skip update if it would be a no-op */
+	if (!!(priv->isolated_ports & BIT_ULL(port)) == isolate) {
+		pr_debug("%s: no-op\n", __func__);
+		goto out_unlock;
+	}
```
[the remark this answers is not in the document]
Comment (sudoBash418) [reply target/linux/realtek/files-5.15/drivers/net/dsa/rtl83xx/dsa.c:1287]: Only in terms of performance - though I haven't actually benchmarked it, so I'm not sure if it actually matters.

File: target/linux/realtek/patches-5.15/600-net-dsa-allow-offloading-bridge-port-isolation.patch [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+--- a/net/bridge/br_switchdev.c
++++ b/net/bridge/br_switchdev.c
+@@ -71,7 +71,8 @@ bool nbp_switchdev_allowed_egress(const
+ 
+ /* Flags that can be offloaded to hardware */
+ #define BR_PORT_FLAGS_HW_OFFLOAD (BR_LEARNING | BR_FLOOD | \
+-				  BR_MCAST_FLOOD | BR_BCAST_FLOOD)
++                                  BR_MCAST_FLOOD | BR_BCAST_FLOOD | \
++                                  BR_ISOLATED)
```
[the remark this answers is not in the document]
Comment (sudoBash418) [reply target/linux/realtek/patches-5.15/600-net-dsa-allow-offloading-bridge-port-isolation.patch:23]: I hadn't thought to check if a similar change was already upstreamed, thanks for the hint!

Unfortunately it looks like commit `c3976a3f8445` depends on a previous commit from v5.18, `fa1c83342987`, which adds the `BR_PORT_LOCKED` flag.

The `BR_PORT_LOCKED` flag was introduced in v5.18 as well (`a21d9a670d811`: net: bridge: Add support for offloading of locked port flag), which seems a little heavy for a two-line change.

So I'm thinking it's probably not worth backporting? But let me know what you think.

File: target/linux/realtek/patches-5.15/600-net-dsa-allow-offloading-bridge-port-isolation.patch [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+ int br_switchdev_set_port_flag(struct net_bridge_port *p,
+ 			       unsigned long flags,
+--- a/net/dsa/port.c
++++ b/net/dsa/port.c
+@@ -195,7 +195,7 @@ static int dsa_port_inherit_brport_flags
+ 					 struct netlink_ext_ack *extack)
+ {
+ 	const unsigned long mask = BR_LEARNING | BR_FLOOD | BR_MCAST_FLOOD |
+-				   BR_BCAST_FLOOD;
++	                           BR_BCAST_FLOOD | BR_ISOLATED;
```
[the remark this answers is not in the document]
Comment (sudoBash418) [reply target/linux/realtek/patches-5.15/600-net-dsa-allow-offloading-bridge-port-isolation.patch:34]: Splitting the patch makes sense to me.

Regarding upstreaming: I'm not familiar with the procedure there, but I don't see why not.

File: target/linux/realtek/files-5.15/drivers/net/dsa/rtl83xx/dsa.c [hunk 5]
```diff
@@ -1260,6 +1265,44 @@ static void rtl83xx_port_bridge_leave(struct dsa_switch *ds, int port,
...
+	/* update switch isolation bitmask */
+	if (isolate)
+		priv->isolated_ports |= BIT_ULL(port);
+	else
+		priv->isolated_ports &= ~BIT_ULL(port);
+
+	/* if this port is disabled, all traffic should already be blocked unconditionally */
+	if (priv->ports[port].enable)
+		/* enable/disable traffic between ourselves and any isolated neighbors */
+		rtl83xx_port_update_traffic(priv, port);
```
[the remark this answers is not in the document]
Comment (sudoBash418) [reply target/linux/realtek/files-5.15/drivers/net/dsa/rtl83xx/dsa.c:1298]: I'm not sure I follow? If the isolation change was a no-op, then by definition the forwarding matrix shouldn't need to change in response?
(`rtl83xx_port_update_traffic` should be called whenever any of its "inputs" are changed: `enable`, `bridge_neighbors`, `isolated_ports`, etc.)

**Comment by sudoBash418** [comment]
Just pushed these changes:
- fix the bitmask issue
- split the kernel patch into two
- fix accidentally setting `BR_ISOLATED` in `dsa_port_clear_brport_flags`

**Comment by svanheule** [comment]
I've asked for clarification on the submission guidelines with the rest of the team, because the CI is complaining for good reason:
https://github.com/openwrt/actions-shared-workflows/commit/12d9551f2d07ec34ac813da8612c8014fb393af6

File: target/linux/realtek/files-5.15/drivers/net/dsa/rtl83xx/dsa.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (svanheule) [reply target/linux/realtek/files-5.15/drivers/net/dsa/rtl83xx/dsa.c:1287]: It's a nice catch and might save some time e.g. during init. Marginal gains, I suppose :)

File: target/linux/realtek/files-5.15/drivers/net/dsa/rtl83xx/dsa.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (svanheule) [reply target/linux/realtek/files-5.15/drivers/net/dsa/rtl83xx/dsa.c:1298]: If all cases are covered it's fine. Probably better this way than to bluntly performing updates with every change and not catching bugs early on.

File: target/linux/realtek/patches-5.15/600-net-dsa-allow-offloading-bridge-port-isolation.patch (hunk 4 above)
[the remark this answers is not in the document]
Comment (svanheule) [reply target/linux/realtek/patches-5.15/600-net-dsa-allow-offloading-bridge-port-isolation.patch]: Thanks for splitting the patch. With the different descriptions it even makes more sense.

Submitting upstream is a bit more involved than creating a Github account:
https://www.kernel.org/doc/html/latest/process/submitting-patches.html

Out policy is normally to minimize the number of out-of-tree kernel changes. That means submitting changes upstream so they obsolete local patches on kernel upgrades. Be sure to check out the netdev mailing list. If you do it wrong, people will tell you very quickly ;)

[the remark this answers is not in the document]
Comment (svanheule) [reply target/linux/realtek/patches-5.15/600-net-dsa-allow-offloading-bridge-port-isolation.patch]: netdev has it own set of rules:
https://www.kernel.org/doc/html/latest/process/maintainer-netdev.html#netdev-faq

File: target/linux/realtek/patches-5.15/600-net-dsa-allow-offloading-bridge-port-isolation.patch (hunk 3 above)
[the remark this answers is not in the document]
Comment (svanheule) [reply target/linux/realtek/patches-5.15/600-net-dsa-allow-offloading-bridge-port-isolation.patch]: You could take the upstream patch and amend it to remove `BR_PORT_LOCKED`, i.e. rebase the patch on top of 5.15.

**Comment by sudoBash418** [comment]
That is unfortunate. I am not comfortable using my real name, so I guess this PR is blocked by the real-name policy.

Regardless, thank for your time reviewing this PR.

I'll close it for now to reduce noise - hopefully I'll be able to reopen it in the future :slightly_smiling_face:.

**Pull request closed without merging**